src/cryptonote_protocol/quorumnet.cpp

```diff
@@ -0,0 +1,1392 @@
+        auto bit = snw.blinks.find(blink_height);
+        if (bit != snw.blinks.end()) {
+            auto it = bit->second.find(tx_hash);
+            if (it != bit->second.end()) {
+                btxptr = it->second.btxptr;
+                reply_tag = it->second.reply_tag;
+                reply_pubkey = it->second.reply_pubkey;
+            }
+        }
+    }
+
+    if (btxptr)
+        MINFO("Found blink tx in local blink cache");
+    else {
+        MINFO("Blink tx not found in local blink cache; delaying signature verification");
+        std::unique_lock<std::shared_timed_mutex> lock{snw.mutex};
+        auto &delayed = snw.blinks[blink_height][tx_hash].pending_sigs;
+        for (auto &sig : signatures)
+            delayed.insert(std::move(sig));
+        return;
+    }
+
+    process_blink_signatures(snw, *btxptr, blink_quorums, checksum, std::move(signatures), reply_tag, reply_pubkey, m.pubkey);
+}
+
+
+// tag -> {hash, promise, expiry}
+using blink_response = std::pair<cryptonote::blink_result, std::string>;
+struct blink_result_data {
+    crypto::hash hash;
+    std::promise<blink_response> promise;
+    std::chrono::high_resolution_clock::time_point expiry;
+    int remote_count;
+    std::atomic<int> nostart_count{0};
+    std::atomic<int> bad_count{0};
+    std::atomic<int> good_count{0};
+};
+std::unordered_map<uint64_t, blink_result_data> pending_blink_results;
+std::shared_timed_mutex pending_blink_result_mutex;
+
+// Sanity check against runaway active pending blink submissions
+constexpr size_t MAX_ACTIVE_PROMISES = 1000;
+
+std::future<std::pair<cryptonote::blink_result, std::string>> send_blink(void *obj, const std::string &tx_blob) {
+    std::promise<std::pair<cryptonote::blink_result, std::string>> promise;
+    auto future = promise.get_future();
+    cryptonote::transaction tx;
+    crypto::hash tx_hash;
+
+    thread_local std::mt19937_64 rng{std::random_device{}()};
+    uint64_t blink_tag = 0;
+    blink_result_data *brd = nullptr;
+
+    if (!cryptonote::parse_and_validate_tx_from_blob(tx_blob, tx, tx_hash)) {
+        promise.set_value(std::make_pair(cryptonote::blink_result::rejected, "Could not parse transaction data"));
+    } else {
+        auto now = std::chrono::high_resolution_clock::now();
+        bool found = false;
+        std::unique_lock<std::shared_timed_mutex> lock{pending_blink_result_mutex};
+        for (auto it = pending_blink_results.begin(); it != pending_blink_results.end(); ) {
+            auto &brd = it->second;
+            if (brd.expiry >= now) {
+                try { brd.promise.set_value(std::make_pair(cryptonote::blink_result::timeout, "Blink quorum timeout")); }
+                catch (const std::future_error &) { /* ignore */ }
+                it = pending_blink_results.erase(it);
+            } else {
+                if (!found && brd.hash == tx_hash)
+                    found = true;
+                ++it;
+            }
+        }
+        if (found) {
```

Comment:
We need to complete the promise because the rpc connection that submitted it is waiting for it to respond, but we may never get that response (network connectivity issues, or 4+ SNs in one of the two quorums is down) and don't want it to hang indefinitely.

It doesn't stop the blink tx from being processed by the network, however: once you get it out to just one SN you're pretty much guaranteed that it's going to the network (assuming it gets approved and there aren't too many unreachable SNs).  If some timeout here occurs with your local node you may see the timeout but then still see the transaction appear on the network anyway.

I'll poke into the conditions above and either make sure we can't double-set, or that we ignore a throw from trying.

Reply:
Oh I see, it's `brd.promise.set_value` and `promise.set_value` 2 different instances. So doesn't look like it can double set the same promise.